quaint/src/connector/postgres/native/cache.rs

```diff
@@ -0,0 +1,388 @@
+use std::{
+    hash::{BuildHasher, Hash, RandomState},
+    sync::Arc,
+};
+
+use async_trait::async_trait;
+use futures::lock::Mutex;
+use lru_cache::LruCache;
+use postgres_types::Type;
+use tokio_postgres::{Client, Error, Statement};
+
+use crate::connector::metrics::strip_query_traceparent;
+
+use super::query::{PreparedQuery, QueryMetadata, TypedQuery};
+
+/// Types that can be used as a cache for prepared queries and statements.
+#[async_trait]
+pub trait QueryCache: From<CacheSettings> + Send + Sync {
+    /// The type that is returned when a prepared query is requested from the cache.
+    type Query<'a>: PreparedQuery;
+
+    /// Retrieve a prepared query.
+    async fn get_query<'a>(&self, client: &Client, sql: &'a str, types: &[Type]) -> Result<Self::Query<'a>, Error>;
+
+    /// Retrieve a prepared statement.
+    ///
+    /// This is useful in scenarios that require direct access to a prepared statement,
+    /// e.g. describing a query.
+    async fn get_statement(&self, client: &Client, sql: &str, types: &[Type]) -> Result<Statement, Error>;
+}
+
+/// A no-op cache that creates a new prepared statement for every requested query.
+/// Useful when we don't need caching.
+#[derive(Debug, Default)]
+pub struct NoOpCache;
+
+#[async_trait]
+impl QueryCache for NoOpCache {
+    type Query<'a> = Statement;
+
+    #[inline]
+    async fn get_query<'a>(&self, client: &Client, sql: &'a str, types: &[Type]) -> Result<Statement, Error> {
+        self.get_statement(client, sql, types).await
+    }
```

Comment:
Is it possible to get rid of Result? It seems to me that, so far, you only use the Ok() constructor or .unwrap() the result.

```suggestion
    async fn get_query<'a>(&self, client: &Client, sql: &'a str, types: &[Type]) -> Statement {
        self.get_statement(client, sql, types).await
    }
```

Reply:
Nevermind, I replied to myself in https://github.com/prisma/prisma-engines/pull/5082/files#r1881880519.

Reply:
same as above